File: DecisionTrees.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Wed Mar 25 12:21:36 2020

@author: guillem
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# ARBOL DE REGRESIÓN

# Importamos los datos
df = pd.read_csv("/home/guillem/CursosML/machinelearning-az-master/datasets/Part 2 - Regression/Section 8 - Decision Tree Regression/Position_Salaries.csv")
X = df.iloc[:, 1:2]
y = df[['Salary']]

# Ajustamos el modelo de Arboles

from sklearn.tree import DecisionTreeRegressor
regtree = DecisionTreeRegressor(random_state = 0)

regtree.fit(X, y)

# Predecimos una muestra concreta
y_pred = regtree.predict(np.array(6.5).reshape(-1, 1))

# Visualizamos los resultados del Modelo De Arbol
fig, ax = plt.subplots(1, 1)
ax.scatter(X, y, color = "red")
ax.plot(X, regtree.predict(X), color = "blue")
plt.show()

# Visualizamos los resultados más "suavizado"
X_grid = np.arange(0, 10, 0.1)
X_grid = X_grid.reshape(len(X_grid), 1)
fig, ax = plt.subplots(1, 1)
ax.scatter(X, y, color = "red")
ax.plot(X_grid, regtree.predict(X_grid), color = "blue")
plt.show()


# BOSQUES ALEATORIOS

df = pd.read_csv("/home/guillem/CursosML/machinelearning-az-master/datasets/Part 2 - Regression/Section 9 - Random Forest Regression/Position_Salaries.csv")
X = df.iloc[:, 1:2]
y = df[['Salary']]

# Ajustamos el modelo Random Forest
from sklearn.ensemble import RandomForestRegressor
regrf = RandomForestRegressor(n_estimators = 500, random_state = 0)
regrf.fit(X, y)

# Predicción puntual

y_pred = regrf.predict(np.array(6.5).reshape(-1, 1))

# Visualizamos los resultados más "suavizado"

X_grid = np.arange(0, 10, 0.1)
X_grid = X_grid.reshape(len(X_grid), 1)
fig, ax = plt.subplots(1, 1, figsize = (15, 10))
ax.scatter(X, y, color = "red")
ax.plot(X_grid, regrf.predict(X_grid), color = "blue")
ax.annotate('Predicción', xy = (6.5, regrf.predict(np.array(6.5).reshape(-1, 1))),
            xytext = (6.5, 60000), arrowprops = dict(width = 0.1,facecolor = "black"))
plt.show()

File: DecisionTrees.R
# ÁRBOLES DE DECISIÓN

# Cargamos el dataset

df <- read.csv("~/CursosML/machinelearning-az-master/datasets/Part 2 - Regression/Section 8 - Decision Tree Regression/Position_Salaries.csv")

# Ajustamos el Modelo de Arbol de Decision
library(rpart)
regtree <- rpart(Salary ~ Level,
                 data = df)

# Predecimos una muestra concreta
predict(regtree, newdata = data.frame(Level = 6.5))

# Visualizamos el por qué de que vaya mal
library(ggplot2)
ggplot(data = df)+
  geom_point(aes(Level, Salary), colour = "red")+
  geom_line(aes(Level, predict(regtree, newdata = data.frame(Level = Level))), colour = "blue")

# Vamos a hacer un poco de hyperparameter tunning para solcionarlo

regtree <- rpart(Salary ~ Level,
                 data = df,
                 control = rpart.control(minsplit  = 1))

library(ggplot2)
ggplot(data = df)+
  geom_point(aes(Level, Salary), colour = "red")+
  geom_line(aes(Level, predict(regtree, newdata = data.frame(Level = Level))), colour = "blue")

x_grid <- seq(min(df$Level), max(df$Level), 0.1)

ggplot()+
  geom_point(aes(df$Level, df$Salary), colour = "red")+
  geom_line(aes(x_grid, predict(regtree, newdata = data.frame(Level = x_grid))), colour = "blue")

# BOSQUES ALEATORIOS

df <- read.csv("~/CursosML/machinelearning-az-master/datasets/Part 2 - Regression/Section 9 - Random Forest Regression/Position_Salaries.csv")

# Ajustamos el modelo de random forest

library(randomForest)
regrf <- randomForest(x = df[2], 
                      y = df$Salary,
                      ntree = 5000)
summary(regrf)

# Visualizamos los datos de forma ("suavizada")
library(ggplot2)
x_grid <- seq(min(df$Level), max(df$Level), 0.1)

ggplot()+
  geom_point(aes(df$Level, df$Salary), colour = "red")+
  geom_line(aes(x_grid, predict(regrf, newdata = data.frame(Level = x_grid))), colour = "blue")

